schism/compiler.ss

```diff
@@ -490,15 +494,24 @@
   (define (parse-body body env)
     (unless (pair? body) (error 'parse-body "Empty body"))
     (parse-begin (parse-expr (car body) env) (cdr body) env))
-  (define (parse-function function env)
-    (let* ((name (definition-name function))
-           (args (function-formals function))
+  (define (define-function def env)
+    (let* ((name (definition-name def))
+           (args (function-formals def))
            (args* (map rename-var args))
-           (body (parse-body (function-body function)
+           (body (parse-body (function-body def)
                              (add-lexicals args args* env))))
       `(,(cons name args*) ,body)))
-  (define (parse-functions definitions env)
-    (map (lambda (fn) (parse-function fn env)) definitions))
+  (define (define-value def env)
+    (let ((body (if (null? (cddr def))
+                    `(const ,(void))
+                    (parse-expr (caddr def) env))))
+      `(define ,(definition-name def) ,body)))
+  (define (parse-definition def env)
+    (if (function-definition? def)
+        (define-function def env)
+        (define-value def env)))
```

Comment:
This is the line where I'd deal with the environment problem for top-level values. Basically, for the call to `define-function`, we want to pass the whole top level environment. For the call to `define-value`, we need to pass the environment only up to the current definition. Then any unbound variables in the right hand side expression will trigger errors.

Reply:
This makes a lot of sense! The change I just pushed works as expected, but calling `memp` is quadratic. I think `parse-definitions` can be reworked to avoid this